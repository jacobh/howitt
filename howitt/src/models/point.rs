use chrono::{DateTime, TimeZone, Utc};
use itertools::Itertools;
use serde::{Deserialize, Serialize};

use super::ModelId;

pub trait Point: std::fmt::Debug + Clone {
    fn as_geo_point(&self) -> &geo::Point;
    fn elevation_meters(&self) -> Option<&f64>;
    fn to_elevation_point(&self) -> Option<ElevationPoint>;
}

impl Point for geo::Point {
    fn as_geo_point(&self) -> &geo::Point {
        self
    }

    fn elevation_meters(&self) -> Option<&f64> {
        None
    }

    fn to_elevation_point(&self) -> Option<ElevationPoint> {
        None
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct ElevationPoint {
    pub point: geo::Point,
    pub elevation: f64,
}

impl Serialize for ElevationPoint {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        (self.point.x(), self.point.y(), self.elevation).serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for ElevationPoint {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let (x, y, elevation) = Deserialize::deserialize(deserializer)?;
        Ok(ElevationPoint {
            point: geo::Point::new(x, y),
            elevation,
        })
    }
}

impl Point for ElevationPoint {
    fn as_geo_point(&self) -> &geo::Point {
        &self.point
    }

    fn elevation_meters(&self) -> Option<&f64> {
        Some(&self.elevation)
    }

    fn to_elevation_point(&self) -> Option<ElevationPoint> {
        Some(self.clone())
    }
}

impl<'a> Point for &'a ElevationPoint {
    fn as_geo_point(&self) -> &geo::Point {
        &self.point
    }

    fn elevation_meters(&self) -> Option<&f64> {
        Some(&self.elevation)
    }

    fn to_elevation_point(&self) -> Option<ElevationPoint> {
        Some((*self).clone())
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct TemporalElevationPoint {
    pub datetime: DateTime<Utc>,
    pub point: geo::Point,
    pub elevation: f64,
}

impl Serialize for TemporalElevationPoint {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        (
            self.datetime.timestamp(),
            self.point.x(),
            self.point.y(),
            self.elevation,
        )
            .serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for TemporalElevationPoint {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let (timestamp, x, y, elevation) = Deserialize::deserialize(deserializer)?;
        Ok(TemporalElevationPoint {
            datetime: Utc
                .timestamp_opt(timestamp, 0)
                .single()
                .ok_or_else(|| serde::de::Error::custom("couldnt parse timestamp"))?,
            point: geo::Point::new(x, y),
            elevation,
        })
    }
}

impl Point for TemporalElevationPoint {
    fn as_geo_point(&self) -> &geo::Point {
        &self.point
    }

    fn elevation_meters(&self) -> Option<&f64> {
        Some(&self.elevation)
    }

    fn to_elevation_point(&self) -> Option<ElevationPoint> {
        Some(ElevationPoint {
            point: self.point,
            elevation: self.elevation,
        })
    }
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct PointChunk<ID, P> {
    pub model_id: ID,
    pub idx: usize,
    pub points: Vec<P>,
}
impl<ID, P> PointChunk<ID, P>
where
    ID: ModelId,
{
    pub fn new_chunks(model_id: ID, points: impl IntoIterator<Item = P>) -> Vec<PointChunk<ID, P>> {
        points
            .into_iter()
            .chunks(2500)
            .into_iter()
            .enumerate()
            .map(|(idx, points)| PointChunk {
                model_id,
                idx,
                points: points.collect(),
            })
            .collect()
    }
    pub fn iter_points(chunks: &[PointChunk<ID, P>]) -> impl Iterator<Item = &P> + '_ {
        chunks.iter().flat_map(|chunk| chunk.points.iter())
    }
}
