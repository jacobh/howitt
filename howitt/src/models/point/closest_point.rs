use super::{point::Point, point_delta::PointDelta};

pub fn closest_point<'a, P: Point>(
    point: &P,
    points: impl Iterator<Item = &'a P>,
) -> Option<(&'a P, PointDelta<P::DeltaData>)> {
    points
        .map(|p| {
            let delta = PointDelta::from_points(point, p);
            (p, delta)
        })
        .min_by_key(|(_, delta)| delta.clone())
}
