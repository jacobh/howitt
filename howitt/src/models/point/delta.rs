use crate::ext::iter::ScanAllExt;
use geo::{Bearing, Distance, Haversine};
use itertools::Itertools;

use super::{Point, WithDatetime, WithElevation};

pub trait Delta<P> {
    fn delta(value1: &P, value2: &P) -> Self;
}

pub trait AccumulatingDelta<P>: Sized {
    fn running_totals(values: &[P]) -> Vec<Self>;
}

// ----

#[derive(Debug)]
pub struct DistanceDelta(pub f64);
impl<P: Point> Delta<P> for DistanceDelta {
    fn delta(value1: &P, value2: &P) -> Self {
        let distance = Haversine::distance(*value1.as_geo_point(), *value2.as_geo_point());
        DistanceDelta(distance)
    }
}

impl<P: Point> AccumulatingDelta<P> for DistanceDelta {
    fn running_totals(values: &[P]) -> Vec<Self> {
        std::iter::once(DistanceDelta(0.0))
            .chain(
                values
                    .iter()
                    .tuple_windows()
                    .map(|(a, b)| Self::delta(a, b)),
            )
            .scan_all(0.0, |acc, DistanceDelta(d)| {
                *acc += d;
                DistanceDelta(*acc)
            })
            .collect()
    }
}

#[derive(Debug)]
pub struct BearingDelta(pub f64);
impl<P: Point> Delta<P> for BearingDelta {
    fn delta(value1: &P, value2: &P) -> Self {
        let bearing = Haversine::bearing(*value1.as_geo_point(), *value2.as_geo_point());
        BearingDelta(bearing)
    }
}

#[derive(Debug)]
pub struct ElevationDelta(pub f64);
impl<P: WithElevation> Delta<P> for ElevationDelta {
    fn delta(value1: &P, value2: &P) -> Self {
        // Simple elevation difference
        let delta = value2.elevation() - value1.elevation();
        ElevationDelta(delta)
    }
}

#[derive(Debug)]
pub struct ElevationGainDelta(pub f64);

impl<P: WithElevation> AccumulatingDelta<P> for ElevationGainDelta {
    fn running_totals(values: &[P]) -> Vec<Self> {
        std::iter::once(ElevationGainDelta(0.0))
            .chain(values.iter().tuple_windows().map(|(a, b)| {
                let delta = b.elevation() - a.elevation();
                ElevationGainDelta(if delta > 0.0 { delta } else { 0.0 })
            }))
            .scan_all(0.0, |acc, ElevationGainDelta(d)| {
                *acc += d;
                ElevationGainDelta(*acc)
            })
            .collect()
    }
}

#[derive(Debug)]
pub struct ElevationLossDelta(pub f64);

impl<P: WithElevation> AccumulatingDelta<P> for ElevationLossDelta {
    fn running_totals(values: &[P]) -> Vec<Self> {
        std::iter::once(ElevationLossDelta(0.0))
            .chain(values.iter().tuple_windows().map(|(a, b)| {
                let delta = b.elevation() - a.elevation();
                ElevationLossDelta(if delta < 0.0 { -delta } else { 0.0 })
            }))
            .scan_all(0.0, |acc, ElevationLossDelta(d)| {
                *acc += d;
                ElevationLossDelta(*acc)
            })
            .collect()
    }
}

#[derive(Debug)]
pub struct ElapsedDelta(pub chrono::Duration);
impl<P: WithDatetime> Delta<P> for ElapsedDelta {
    fn delta(value1: &P, value2: &P) -> Self {
        // Time difference between points
        let delta = value2.datetime().signed_duration_since(*value1.datetime());
        ElapsedDelta(delta)
    }
}

impl<P: WithDatetime> AccumulatingDelta<P> for ElapsedDelta {
    fn running_totals(values: &[P]) -> Vec<Self> {
        std::iter::once(ElapsedDelta(chrono::Duration::zero()))
            .chain(
                values
                    .iter()
                    .tuple_windows()
                    .map(|(a, b)| Self::delta(a, b)),
            )
            .scan_all(chrono::Duration::zero(), |acc, ElapsedDelta(e)| {
                *acc = *acc + e;
                ElapsedDelta(*acc)
            })
            .collect()
    }
}

// ----------

impl<P, T1> Delta<P> for (T1,)
where
    T1: Delta<P>,
{
    fn delta(value1: &P, value2: &P) -> Self {
        (T1::delta(value1, value2),)
    }
}

impl<P, T1, T2> Delta<P> for (T1, T2)
where
    T1: Delta<P>,
    T2: Delta<P>,
{
    fn delta(value1: &P, value2: &P) -> Self {
        (T1::delta(value1, value2), T2::delta(value1, value2))
    }
}

impl<P, T1, T2, T3> Delta<P> for (T1, T2, T3)
where
    T1: Delta<P>,
    T2: Delta<P>,
    T3: Delta<P>,
{
    fn delta(value1: &P, value2: &P) -> Self {
        (
            T1::delta(value1, value2),
            T2::delta(value1, value2),
            T3::delta(value1, value2),
        )
    }
}

impl<P, T1, T2, T3, T4> Delta<P> for (T1, T2, T3, T4)
where
    T1: Delta<P>,
    T2: Delta<P>,
    T3: Delta<P>,
    T4: Delta<P>,
{
    fn delta(value1: &P, value2: &P) -> Self {
        (
            T1::delta(value1, value2),
            T2::delta(value1, value2),
            T3::delta(value1, value2),
            T4::delta(value1, value2),
        )
    }
}

impl<P, T1, T2, T3, T4, T5> Delta<P> for (T1, T2, T3, T4, T5)
where
    T1: Delta<P>,
    T2: Delta<P>,
    T3: Delta<P>,
    T4: Delta<P>,
    T5: Delta<P>,
{
    fn delta(value1: &P, value2: &P) -> Self {
        (
            T1::delta(value1, value2),
            T2::delta(value1, value2),
            T3::delta(value1, value2),
            T4::delta(value1, value2),
            T5::delta(value1, value2),
        )
    }
}

impl<P, T1> AccumulatingDelta<P> for (T1,)
where
    T1: AccumulatingDelta<P>,
{
    fn running_totals(values: &[P]) -> Vec<Self> {
        T1::running_totals(values)
            .into_iter()
            .map(|t1| (t1,))
            .collect()
    }
}

impl<P, T1, T2> AccumulatingDelta<P> for (T1, T2)
where
    T1: AccumulatingDelta<P>,
    T2: AccumulatingDelta<P>,
{
    fn running_totals(values: &[P]) -> Vec<Self> {
        let t1_deltas = T1::running_totals(values);
        let t2_deltas = T2::running_totals(values);
        t1_deltas.into_iter().zip(t2_deltas).collect()
    }
}

impl<P, T1, T2, T3> AccumulatingDelta<P> for (T1, T2, T3)
where
    T1: AccumulatingDelta<P>,
    T2: AccumulatingDelta<P>,
    T3: AccumulatingDelta<P>,
{
    fn running_totals(values: &[P]) -> Vec<Self> {
        let t1_deltas = T1::running_totals(values);
        let t2_deltas = T2::running_totals(values);
        let t3_deltas = T3::running_totals(values);
        t1_deltas
            .into_iter()
            .zip(t2_deltas)
            .zip(t3_deltas)
            .map(|((t1, t2), t3)| (t1, t2, t3))
            .collect()
    }
}

impl<P, T1, T2, T3, T4> AccumulatingDelta<P> for (T1, T2, T3, T4)
where
    T1: AccumulatingDelta<P>,
    T2: AccumulatingDelta<P>,
    T3: AccumulatingDelta<P>,
    T4: AccumulatingDelta<P>,
{
    fn running_totals(values: &[P]) -> Vec<Self> {
        let t1_deltas = T1::running_totals(values);
        let t2_deltas = T2::running_totals(values);
        let t3_deltas = T3::running_totals(values);
        let t4_deltas = T4::running_totals(values);
        t1_deltas
            .into_iter()
            .zip(t2_deltas)
            .zip(t3_deltas)
            .zip(t4_deltas)
            .map(|(((t1, t2), t3), t4)| (t1, t2, t3, t4))
            .collect()
    }
}

impl<P, T1, T2, T3, T4, T5> AccumulatingDelta<P> for (T1, T2, T3, T4, T5)
where
    T1: AccumulatingDelta<P>,
    T2: AccumulatingDelta<P>,
    T3: AccumulatingDelta<P>,
    T4: AccumulatingDelta<P>,
    T5: AccumulatingDelta<P>,
{
    fn running_totals(values: &[P]) -> Vec<Self> {
        let t1_deltas = T1::running_totals(values);
        let t2_deltas = T2::running_totals(values);
        let t3_deltas = T3::running_totals(values);
        let t4_deltas = T4::running_totals(values);
        let t5_deltas = T5::running_totals(values);
        t1_deltas
            .into_iter()
            .zip(t2_deltas)
            .zip(t3_deltas)
            .zip(t4_deltas)
            .zip(t5_deltas)
            .map(|((((t1, t2), t3), t4), t5)| (t1, t2, t3, t4, t5))
            .collect()
    }
}

// ... existing code ...

#[cfg(test)]
mod tests {
    use chrono::{TimeZone, Utc};
    use geo::Point as GeoPoint;

    use super::*;
    use crate::models::point::{
        elevation_point::ElevationPoint, temporal_elevation_point::TemporalElevationPoint,
    };

    #[test]
    fn test_distance_delta() {
        let points = vec![
            ElevationPoint {
                point: GeoPoint::new(145.0, -37.0),
                elevation: 100.0,
            },
            ElevationPoint {
                point: GeoPoint::new(145.1, -37.1),
                elevation: 200.0,
            },
            ElevationPoint {
                point: GeoPoint::new(145.2, -37.2),
                elevation: 300.0,
            },
        ];

        let deltas = DistanceDelta::running_totals(&points);
        insta::assert_debug_snapshot!((points, deltas));
    }

    #[test]
    fn test_elevation_gain_loss_delta() {
        let points = vec![
            ElevationPoint {
                point: GeoPoint::new(145.0, -37.0),
                elevation: 100.0,
            },
            ElevationPoint {
                point: GeoPoint::new(145.1, -37.1),
                elevation: 200.0,
            },
            ElevationPoint {
                point: GeoPoint::new(145.2, -37.2),
                elevation: 150.0,
            },
            ElevationPoint {
                point: GeoPoint::new(145.3, -37.3),
                elevation: 300.0,
            },
        ];

        let gains = ElevationGainDelta::running_totals(&points);
        let losses = ElevationLossDelta::running_totals(&points);

        insta::assert_debug_snapshot!((points, gains, losses));
    }

    #[test]
    fn test_temporal_deltas() {
        let points = vec![
            TemporalElevationPoint {
                point: GeoPoint::new(145.0, -37.0),
                elevation: 100.0,
                datetime: Utc.timestamp_opt(1000, 0).unwrap(),
            },
            TemporalElevationPoint {
                point: GeoPoint::new(145.1, -37.1),
                elevation: 200.0,
                datetime: Utc.timestamp_opt(2000, 0).unwrap(),
            },
            TemporalElevationPoint {
                point: GeoPoint::new(145.2, -37.2),
                elevation: 150.0,
                datetime: Utc.timestamp_opt(3500, 0).unwrap(),
            },
        ];

        let deltas = <(ElapsedDelta, DistanceDelta, ElevationGainDelta)>::running_totals(&points);
        insta::assert_debug_snapshot!((points, deltas));
    }

    #[test]
    fn test_single_point_delta() {
        let p1 = ElevationPoint {
            point: GeoPoint::new(145.0, -37.0),
            elevation: 100.0,
        };
        let p2 = ElevationPoint {
            point: GeoPoint::new(145.1, -37.1),
            elevation: 200.0,
        };

        let distance = DistanceDelta::delta(&p1, &p2);
        let elevation = ElevationDelta::delta(&p1, &p2);
        let bearing = BearingDelta::delta(&p1, &p2);

        insta::assert_debug_snapshot!(((p1, p2), distance, elevation, bearing));
    }
}
