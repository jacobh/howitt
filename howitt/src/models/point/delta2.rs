use geo::{Bearing, Distance, Haversine};

use super::{Point, WithDatetime, WithElevation};

pub trait Delta<P>: Sized {
    fn delta(value1: &P, value2: &P) -> Self;
}

pub trait CumulativeDelta<P>: Delta<P> {
    fn cumulative_deltas(values: &[P]) -> Vec<Self>;
}

pub struct DistanceDelta(pub f64);
impl<P: Point> Delta<P> for DistanceDelta {
    fn delta(value1: &P, value2: &P) -> Self {
        let distance = Haversine::distance(*value1.as_geo_point(), *value2.as_geo_point());
        DistanceDelta(distance)
    }
}

impl<P: Point> CumulativeDelta<P> for DistanceDelta {
    fn cumulative_deltas(values: &[P]) -> Vec<Self> {
        std::iter::once(DistanceDelta(0.0))
            .chain(values.windows(2).map(|w| Self::delta(&w[0], &w[1])))
            .scan(0.0, |acc, DistanceDelta(d)| {
                *acc += d;
                Some(DistanceDelta(*acc))
            })
            .collect()
    }
}

pub struct BearingDelta(pub f64);
impl<P: Point> Delta<P> for BearingDelta {
    fn delta(value1: &P, value2: &P) -> Self {
        let bearing = Haversine::bearing(*value1.as_geo_point(), *value2.as_geo_point());
        BearingDelta(bearing)
    }
}

pub struct ElevationDelta(pub f64);
impl<P: WithElevation> Delta<P> for ElevationDelta {
    fn delta(value1: &P, value2: &P) -> Self {
        // Simple elevation difference
        let delta = value2.elevation() - value1.elevation();
        ElevationDelta(delta)
    }
}

impl<P: WithElevation> CumulativeDelta<P> for ElevationDelta {
    fn cumulative_deltas(values: &[P]) -> Vec<Self> {
        std::iter::once(ElevationDelta(0.0))
            .chain(values.windows(2).map(|w| Self::delta(&w[0], &w[1])))
            .scan(0.0, |acc, ElevationDelta(e)| {
                *acc += e;
                Some(ElevationDelta(*acc))
            })
            .collect()
    }
}

pub struct ElapsedDelta(pub chrono::Duration);
impl<P: WithDatetime> Delta<P> for ElapsedDelta {
    fn delta(value1: &P, value2: &P) -> Self {
        // Time difference between points
        let delta = value2.datetime().signed_duration_since(*value1.datetime());
        ElapsedDelta(delta)
    }
}

impl<P: WithDatetime> CumulativeDelta<P> for ElapsedDelta {
    fn cumulative_deltas(values: &[P]) -> Vec<Self> {
        std::iter::once(ElapsedDelta(chrono::Duration::zero()))
            .chain(values.windows(2).map(|w| Self::delta(&w[0], &w[1])))
            .scan(chrono::Duration::zero(), |acc, ElapsedDelta(e)| {
                *acc = *acc + e;
                Some(ElapsedDelta(*acc))
            })
            .collect()
    }
}

// Composite Deltas

pub struct DistanceElevationDelta {
    pub distance: DistanceDelta,
    pub elevation: ElevationDelta,
}

impl<P: Point + WithElevation> Delta<P> for DistanceElevationDelta {
    fn delta(value1: &P, value2: &P) -> Self {
        DistanceElevationDelta {
            distance: DistanceDelta::delta(value1, value2),
            elevation: ElevationDelta::delta(value1, value2),
        }
    }
}

impl<P: Point + WithElevation> CumulativeDelta<P> for DistanceElevationDelta {
    fn cumulative_deltas(values: &[P]) -> Vec<Self> {
        let distances = DistanceDelta::cumulative_deltas(values);
        let elevations = ElevationDelta::cumulative_deltas(values);
        distances
            .into_iter()
            .zip(elevations)
            .map(|(distance, elevation)| DistanceElevationDelta {
                distance,
                elevation,
            })
            .collect()
    }
}

pub struct DistanceElapsedDelta {
    pub distance: DistanceDelta,
    pub elapsed: ElapsedDelta,
}

impl<P: Point + WithDatetime> Delta<P> for DistanceElapsedDelta {
    fn delta(value1: &P, value2: &P) -> Self {
        DistanceElapsedDelta {
            distance: DistanceDelta::delta(value1, value2),
            elapsed: ElapsedDelta::delta(value1, value2),
        }
    }
}

impl<P: Point + WithDatetime> CumulativeDelta<P> for DistanceElapsedDelta {
    fn cumulative_deltas(values: &[P]) -> Vec<Self> {
        let distances = DistanceDelta::cumulative_deltas(values);
        let elapsed = ElapsedDelta::cumulative_deltas(values);
        distances
            .into_iter()
            .zip(elapsed)
            .map(|(distance, elapsed)| DistanceElapsedDelta { distance, elapsed })
            .collect()
    }
}

pub struct DistanceElevationElapsedDelta {
    pub distance: DistanceDelta,
    pub elevation: ElevationDelta,
    pub elapsed: ElapsedDelta,
}

impl<P: Point + WithElevation + WithDatetime> Delta<P> for DistanceElevationElapsedDelta {
    fn delta(value1: &P, value2: &P) -> Self {
        DistanceElevationElapsedDelta {
            distance: DistanceDelta::delta(value1, value2),
            elevation: ElevationDelta::delta(value1, value2),
            elapsed: ElapsedDelta::delta(value1, value2),
        }
    }
}

impl<P: Point + WithElevation + WithDatetime> CumulativeDelta<P> for DistanceElevationElapsedDelta {
    fn cumulative_deltas(values: &[P]) -> Vec<Self> {
        let distances = DistanceDelta::cumulative_deltas(values);
        let elevations = ElevationDelta::cumulative_deltas(values);
        let elapsed = ElapsedDelta::cumulative_deltas(values);
        distances
            .into_iter()
            .zip(elevations)
            .zip(elapsed)
            .map(
                |((distance, elevation), elapsed)| DistanceElevationElapsedDelta {
                    distance,
                    elevation,
                    elapsed,
                },
            )
            .collect()
    }
}
