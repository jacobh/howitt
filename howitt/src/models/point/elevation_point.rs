use serde::{Deserialize, Serialize};

use super::{point::Point, point_delta::ElevationDelta};

#[derive(Debug, PartialEq, Clone)]
pub struct ElevationPoint {
    pub point: geo::Point,
    pub elevation: f64,
}

impl Serialize for ElevationPoint {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        (self.point.x(), self.point.y(), self.elevation).serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for ElevationPoint {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let (x, y, elevation) = Deserialize::deserialize(deserializer)?;
        Ok(ElevationPoint {
            point: geo::Point::new(x, y),
            elevation,
        })
    }
}

impl Point for ElevationPoint {
    type DeltaData = ElevationDelta;

    fn as_geo_point(&self) -> &geo::Point {
        &self.point
    }

    fn elevation_meters(&self) -> Option<&f64> {
        Some(&self.elevation)
    }

    fn to_elevation_point(&self) -> Option<ElevationPoint> {
        Some(self.clone())
    }

    fn delta(&self, other: &Self) -> Self::DeltaData {
        ElevationDelta {
            elevation_gain: other.elevation - self.elevation,
        }
    }
}

impl Point for &ElevationPoint {
    type DeltaData = ElevationDelta;

    fn as_geo_point(&self) -> &geo::Point {
        &self.point
    }

    fn elevation_meters(&self) -> Option<&f64> {
        Some(&self.elevation)
    }

    fn to_elevation_point(&self) -> Option<ElevationPoint> {
        Some((*self).clone())
    }

    fn delta(&self, other: &Self) -> Self::DeltaData {
        ElevationDelta {
            elevation_gain: other.elevation - self.elevation,
        }
    }
}
