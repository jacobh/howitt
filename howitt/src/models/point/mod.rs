pub mod closest_point;
pub mod delta2;
pub mod elevation_point;
pub mod point;
pub mod point_chunk;
pub mod progress;
pub mod temporal_elevation_point;

pub use closest_point::*;
pub use elevation_point::*;
pub use point::*;
pub use point_chunk::*;
pub use temporal_elevation_point::*;
