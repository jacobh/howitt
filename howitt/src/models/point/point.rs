use super::point_delta::DeltaData;

pub trait Point: std::fmt::Debug + Clone {
    type DeltaData: DeltaData;

    fn as_geo_point(&self) -> &geo::Point;
    fn delta(&self, other: &Self) -> Self::DeltaData;

    fn x_y(&self) -> (f64, f64) {
        geo::Point::x_y(*self.as_geo_point())
    }

    fn ordered_x_y(
        &self,
    ) -> (
        ordered_float::OrderedFloat<f64>,
        ordered_float::OrderedFloat<f64>,
    ) {
        let (x, y) = self.x_y();

        (
            ordered_float::OrderedFloat(x),
            ordered_float::OrderedFloat(y),
        )
    }

    fn to_x_y_vec(&self) -> Vec<f64> {
        let (x, y) = self.x_y();
        vec![x, y]
    }

    fn into_x_y_vec(self) -> Vec<f64> {
        let (x, y) = self.x_y();
        vec![x, y]
    }
}

impl Point for geo::Point {
    type DeltaData = ();

    fn as_geo_point(&self) -> &geo::Point {
        self
    }

    fn delta(&self, _: &Self) -> Self::DeltaData {}
}

impl Point for &geo::Point {
    type DeltaData = ();

    fn as_geo_point(&self) -> &geo::Point {
        self
    }

    fn delta(&self, _: &Self) -> Self::DeltaData {}
}
