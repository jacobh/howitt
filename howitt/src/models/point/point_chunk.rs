use itertools::Itertools;
use serde::{Deserialize, Serialize};

use crate::models::ModelId;

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct PointChunk<ID, P> {
    pub model_id: ID,
    pub idx: usize,
    pub points: Vec<P>,
}
impl<ID, P> PointChunk<ID, P>
where
    P: 'static,
    ID: ModelId,
{
    pub fn new_chunks(model_id: ID, points: impl IntoIterator<Item = P>) -> Vec<PointChunk<ID, P>> {
        points
            .into_iter()
            .chunks(2500)
            .into_iter()
            .enumerate()
            .map(|(idx, points)| PointChunk {
                model_id,
                idx,
                points: points.collect(),
            })
            .collect()
    }
    pub fn iter_points(chunks: &[PointChunk<ID, P>]) -> impl Iterator<Item = &P> + '_ {
        chunks.iter().flat_map(|chunk| chunk.points.iter())
    }
    pub fn into_iter_points(chunks: Vec<PointChunk<ID, P>>) -> impl Iterator<Item = P> + 'static {
        chunks
            .into_iter()
            .flat_map(|chunk| chunk.points.into_iter())
    }
}
