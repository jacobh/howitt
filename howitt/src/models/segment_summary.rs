use either::Either;
use serde::{Deserialize, Serialize};

use super::point::Point;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum CardinalDirection {
    North,
    South,
    East,
    West,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum SlopeEnd {
    Uphill,
    Downhill,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Terminus<P: Point> {
    pub direction: CardinalDirection,
    pub slope_end: SlopeEnd,
    pub point: P,
}

pub type Termini<P> = Either<P, (Terminus<P>, Terminus<P>)>;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SegmentSummary<P: Point> {
    pub distance_m: f64,
    pub elevation: Option<ElevationSummary>,
    pub termini: Termini<P>,
}

#[derive(Debug, Default, Clone, Serialize, Deserialize)]
pub struct ElevationSummary {
    pub elevation_ascent_m: f64,
    pub elevation_descent_m: f64,
}
