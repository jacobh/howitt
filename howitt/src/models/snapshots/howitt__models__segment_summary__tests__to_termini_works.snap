---
source: howitt/src/models/segment_summary.rs
expression: result
---
(
    Terminus {
        direction: West,
        point: ElevationPoint {
            point: Point(
                Coord {
                    x: 146.60587,
                    y: -37.2154,
                },
            ),
            elevation: 1100.0,
        },
        distance_from_start: 0.0,
        elevation: Some(
            TerminusElevation {
                slope_end: Downhill,
                elevation_gain_from_start: 0.0,
            },
        ),
    },
    Terminus {
        direction: East,
        point: ElevationPoint {
            point: Point(
                Coord {
                    x: 146.68021,
                    y: -37.20515,
                },
            ),
            elevation: 1400.0,
        },
        distance_from_start: 6696.194096387852,
        elevation: Some(
            TerminusElevation {
                slope_end: Uphill,
                elevation_gain_from_start: 300.0,
            },
        ),
    },
)
